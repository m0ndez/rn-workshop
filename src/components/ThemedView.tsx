import { View, type ViewProps } from "react-native";

import { useThemeColor } from "@/hooks/useThemeColor";
import { forwardRef } from "react";

export type ThemedViewProps = ViewProps & {
  lightColor?: string;
  darkColor?: string;
};

export const ThemedView = forwardRef<View, ThemedViewProps>((props, ref) => {
  const { style, lightColor, darkColor, ...otherProps } = props;

  const backgroundColor = useThemeColor(
    { light: lightColor, dark: darkColor },
    "background"
  );

  return (
    <View ref={ref} style={[{ backgroundColor }, style]} {...otherProps} />
  );
});
