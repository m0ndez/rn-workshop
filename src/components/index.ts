export * from "./navigation";
export * from "./button";
export * from "./surface";
export * from "./status-bar";
export * from "./wave-emoji";
export * from "./collapsible";
export * from "./external-link";
export * from "./parallax-scroll-view";
