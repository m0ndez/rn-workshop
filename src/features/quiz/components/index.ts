export * from "./answer-choices";
export * from "./ended-quiz";
export * from "./quiz-status-bar";
