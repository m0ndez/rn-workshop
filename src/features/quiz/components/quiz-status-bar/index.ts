export * from "./quiz-status-bar";
