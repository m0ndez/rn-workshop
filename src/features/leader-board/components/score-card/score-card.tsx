import { Surface, WaveEmoji } from "@/components";
import { ElementRef, forwardRef, useMemo } from "react";
import { View } from "react-native";
import { Text, useTheme } from "react-native-paper";

interface ScoreCardProps {
  name: string;
  score: number;
  rankPosition: number;
  isYou?: boolean;
}

export const ScoreCard = forwardRef<ElementRef<typeof Surface>, ScoreCardProps>(
  (props, ref) => {
    const { name, score, rankPosition = 1, isYou = false } = props;

    const theme = useTheme();

    const colorRank = useMemo(() => {
      if (rankPosition === 1) {
        return "#FFD700";
      }
      if (rankPosition === 2) {
        return "#C0C0C0";
      }
      if (rankPosition === 3) {
        return "#cd7f32";
      }
      return theme.colors.surface;
    }, [rankPosition]);

    return (
      <Surface
        ref={ref}
        style={{
          flexDirection: "row",
          alignItems: "center",
          justifyContent: "space-between",
          backgroundColor: colorRank,
        }}
      >
        {rankPosition < 3 && (
          <View
            style={{
              position: "absolute",
              top: 0,
              left: -12,
            }}
          >
            <WaveEmoji emoji="🏆" />
          </View>
        )}
        <View
          style={{
            flexDirection: "row",
            gap: 4,
            alignItems: "center",
          }}
        >
          <Text variant="titleMedium">{name}</Text>
          {isYou && <Text variant="labelMedium">(You)</Text>}
        </View>
        <Text variant="labelLarge">{score}</Text>
      </Surface>
    );
  }
);
