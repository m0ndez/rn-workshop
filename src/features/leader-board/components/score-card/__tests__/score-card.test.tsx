import React, { ElementRef } from "react";
import { render } from "@testing-library/react-native";
import { ScoreCard } from "../score-card";
import { Surface } from "@/components";

describe("ScoreCard component", () => {
  it("renders the name, score, and rankPosition 1 correctly", () => {
    const name = "John Doe";
    const score = 100;
    const rankPosition = 1;
    const { getByText, toJSON } = render(
      <ScoreCard name={name} score={score} rankPosition={rankPosition} />
    );
    expect(getByText(name)).toBeTruthy();
    expect(getByText(score.toString())).toBeTruthy();
    expect(getByText(`🏆`)).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });

  it("renders the name, score, and rankPosition 2 correctly", () => {
    const name = "John Doe";
    const score = 100;
    const rankPosition = 2;
    const { getByText, toJSON } = render(
      <ScoreCard name={name} score={score} rankPosition={rankPosition} />
    );
    expect(getByText(name)).toBeTruthy();
    expect(getByText(score.toString())).toBeTruthy();
    expect(getByText(`🏆`)).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });

  it("renders the name, score, and rankPosition 3 correctly", () => {
    const name = "John Doe";
    const score = 100;
    const rankPosition = 3;
    const { getByText, toJSON } = render(
      <ScoreCard name={name} score={score} rankPosition={rankPosition} />
    );
    expect(getByText(name)).toBeTruthy();
    expect(getByText(score.toString())).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });

  it("renders the name, score, and rankPosition 4 correctly", () => {
    const name = "John Doe";
    const score = 100;
    const rankPosition = 4;
    const { getByText, toJSON } = render(
      <ScoreCard name={name} score={score} rankPosition={rankPosition} />
    );
    expect(getByText(name)).toBeTruthy();
    expect(getByText(score.toString())).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });

  it("renders 'You' when isYou prop is true", () => {
    const name = "John Doe";
    const score = 100;
    const rankPosition = 1;

    const { getByText } = render(
      <ScoreCard name={name} score={score} rankPosition={rankPosition} isYou />
    );

    expect(getByText("(You)")).toBeTruthy();
  });

  it("renders without crashing when ref is passed", () => {
    const name = "John Doe";
    const score = 100;
    const rankPosition = 1;

    const ref = React.createRef<ElementRef<typeof Surface>>();

    expect(() =>
      render(
        <ScoreCard
          name={name}
          score={score}
          rankPosition={rankPosition}
          ref={ref}
        />
      )
    ).not.toThrow();
  });
});
