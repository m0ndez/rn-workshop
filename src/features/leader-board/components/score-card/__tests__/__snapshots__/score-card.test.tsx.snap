// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScoreCard component renders the name, score, and rankPosition 1 correctly 1`] = `
<View
  collapsable={false}
  style={
    {
      "backgroundColor": "#FFD700",
      "borderRadius": 8,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 1,
        "width": 0,
      },
      "shadowOpacity": 0.15,
      "shadowRadius": 3,
    }
  }
  testID="surface-outer-layer"
>
  <View
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#FFD700",
        "borderRadius": 8,
        "flex": undefined,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "padding": 16,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.3,
        "shadowRadius": 1,
      }
    }
    testID="surface"
  >
    <View
      style={
        {
          "left": -12,
          "position": "absolute",
          "top": 0,
        }
      }
    >
      <View
        collapsable={false}
        jestAnimatedStyle={
          {
            "value": {
              "transform": [
                {
                  "rotate": "0deg",
                },
              ],
            },
          }
        }
        style={
          {
            "transform": [
              {
                "rotate": "0deg",
              },
            ],
          }
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontWeight": "400",
                "letterSpacing": 0,
              },
              {
                "writingDirection": "ltr",
              },
              {
                "fontSize": 28,
                "lineHeight": 32,
                "marginTop": -6,
              },
            ]
          }
        >
          üèÜ
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "gap": 4,
        }
      }
    >
      <Text
        style={
          [
            {
              "textAlign": "left",
            },
            {
              "color": "rgba(28, 27, 31, 1)",
              "writingDirection": "ltr",
            },
            [
              {
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": "500",
                "letterSpacing": 0.15,
                "lineHeight": 24,
              },
              undefined,
            ],
          ]
        }
      >
        John Doe
      </Text>
    </View>
    <Text
      style={
        [
          {
            "textAlign": "left",
          },
          {
            "color": "rgba(28, 27, 31, 1)",
            "writingDirection": "ltr",
          },
          [
            {
              "fontFamily": "System",
              "fontSize": 14,
              "fontWeight": "500",
              "letterSpacing": 0.1,
              "lineHeight": 20,
            },
            undefined,
          ],
        ]
      }
    >
      100
    </Text>
  </View>
</View>
`;

exports[`ScoreCard component renders the name, score, and rankPosition 2 correctly 1`] = `
<View
  collapsable={false}
  style={
    {
      "backgroundColor": "#C0C0C0",
      "borderRadius": 8,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 1,
        "width": 0,
      },
      "shadowOpacity": 0.15,
      "shadowRadius": 3,
    }
  }
  testID="surface-outer-layer"
>
  <View
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#C0C0C0",
        "borderRadius": 8,
        "flex": undefined,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "padding": 16,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.3,
        "shadowRadius": 1,
      }
    }
    testID="surface"
  >
    <View
      style={
        {
          "left": -12,
          "position": "absolute",
          "top": 0,
        }
      }
    >
      <View
        collapsable={false}
        jestAnimatedStyle={
          {
            "value": {
              "transform": [
                {
                  "rotate": "0deg",
                },
              ],
            },
          }
        }
        style={
          {
            "transform": [
              {
                "rotate": "0deg",
              },
            ],
          }
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontWeight": "400",
                "letterSpacing": 0,
              },
              {
                "writingDirection": "ltr",
              },
              {
                "fontSize": 28,
                "lineHeight": 32,
                "marginTop": -6,
              },
            ]
          }
        >
          üèÜ
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "gap": 4,
        }
      }
    >
      <Text
        style={
          [
            {
              "textAlign": "left",
            },
            {
              "color": "rgba(28, 27, 31, 1)",
              "writingDirection": "ltr",
            },
            [
              {
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": "500",
                "letterSpacing": 0.15,
                "lineHeight": 24,
              },
              undefined,
            ],
          ]
        }
      >
        John Doe
      </Text>
    </View>
    <Text
      style={
        [
          {
            "textAlign": "left",
          },
          {
            "color": "rgba(28, 27, 31, 1)",
            "writingDirection": "ltr",
          },
          [
            {
              "fontFamily": "System",
              "fontSize": 14,
              "fontWeight": "500",
              "letterSpacing": 0.1,
              "lineHeight": 20,
            },
            undefined,
          ],
        ]
      }
    >
      100
    </Text>
  </View>
</View>
`;

exports[`ScoreCard component renders the name, score, and rankPosition 3 correctly 1`] = `
<View
  collapsable={false}
  style={
    {
      "backgroundColor": "#cd7f32",
      "borderRadius": 8,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 1,
        "width": 0,
      },
      "shadowOpacity": 0.15,
      "shadowRadius": 3,
    }
  }
  testID="surface-outer-layer"
>
  <View
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#cd7f32",
        "borderRadius": 8,
        "flex": undefined,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "padding": 16,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.3,
        "shadowRadius": 1,
      }
    }
    testID="surface"
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "gap": 4,
        }
      }
    >
      <Text
        style={
          [
            {
              "textAlign": "left",
            },
            {
              "color": "rgba(28, 27, 31, 1)",
              "writingDirection": "ltr",
            },
            [
              {
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": "500",
                "letterSpacing": 0.15,
                "lineHeight": 24,
              },
              undefined,
            ],
          ]
        }
      >
        John Doe
      </Text>
    </View>
    <Text
      style={
        [
          {
            "textAlign": "left",
          },
          {
            "color": "rgba(28, 27, 31, 1)",
            "writingDirection": "ltr",
          },
          [
            {
              "fontFamily": "System",
              "fontSize": 14,
              "fontWeight": "500",
              "letterSpacing": 0.1,
              "lineHeight": 20,
            },
            undefined,
          ],
        ]
      }
    >
      100
    </Text>
  </View>
</View>
`;

exports[`ScoreCard component renders the name, score, and rankPosition 4 correctly 1`] = `
<View
  collapsable={false}
  style={
    {
      "backgroundColor": "rgba(255, 251, 254, 1)",
      "borderRadius": 8,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 1,
        "width": 0,
      },
      "shadowOpacity": 0.15,
      "shadowRadius": 3,
    }
  }
  testID="surface-outer-layer"
>
  <View
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "rgba(255, 251, 254, 1)",
        "borderRadius": 8,
        "flex": undefined,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "padding": 16,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.3,
        "shadowRadius": 1,
      }
    }
    testID="surface"
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "gap": 4,
        }
      }
    >
      <Text
        style={
          [
            {
              "textAlign": "left",
            },
            {
              "color": "rgba(28, 27, 31, 1)",
              "writingDirection": "ltr",
            },
            [
              {
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": "500",
                "letterSpacing": 0.15,
                "lineHeight": 24,
              },
              undefined,
            ],
          ]
        }
      >
        John Doe
      </Text>
    </View>
    <Text
      style={
        [
          {
            "textAlign": "left",
          },
          {
            "color": "rgba(28, 27, 31, 1)",
            "writingDirection": "ltr",
          },
          [
            {
              "fontFamily": "System",
              "fontSize": 14,
              "fontWeight": "500",
              "letterSpacing": 0.1,
              "lineHeight": 20,
            },
            undefined,
          ],
        ]
      }
    >
      100
    </Text>
  </View>
</View>
`;
